File: src/test/queue/java/QueueTest.java
package queue;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;


public class QueueTest {

	@Test
    public void testQueueCreation() {
        Queue<Integer> queue = new Queue<>(5);
        assertTrue(queue.isEmpty());
        assertEquals(0, queue.size());
    }

    @Test
    public void testAddLast() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertFalse(queue.isEmpty());
        assertEquals(3, queue.size());
        assertEquals(1, queue.getFirst());
        assertEquals(3, queue.getLast());
    }

    @Test
    public void testAddFirst() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addFirst(0);

        assertEquals(3, queue.size());
        assertEquals(0, queue.getFirst());
        assertEquals(2, queue.getLast());
    }

    @Test
    public void testAddAtIndex() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(3);
        queue.add(2, 1);

        assertEquals(3, queue.size());
        assertEquals(1, queue.getFirst());
        assertEquals(3, queue.getLast());
        assertEquals(2, queue.get(1));
    }

    @Test
    public void testRemoveFirst() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(1, queue.removeFirst());
        assertEquals(2, queue.size());
        assertEquals(2, queue.getFirst());
    }

    @Test
    public void testRemoveLast() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(3, queue.removeLast());
        assertEquals(2, queue.size());
        assertEquals(2, queue.getLast());
    }

    @Test
    public void testRemoveAtIndex() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(2, queue.remove(1));
        assertEquals(2, queue.size());
        assertEquals(1, queue.getFirst());
        assertEquals(3, queue.getLast());
    }

    @Test
    public void testIndexOf() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(1, queue.indexOf(2));
        assertEquals(-1, queue.indexOf(4));
    }

    @Test
    public void testLastIndexOf() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(2, queue.lastIndexOf(2));
        assertEquals(-1, queue.lastIndexOf(4));
    }

    @Test
    public void testIsFull() {
        Queue<Integer> queue = new Queue<>(3);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertTrue(queue.isFull());
        assertThrows(IllegalStateException.class, () -> queue.addLast(4));
    }

    @Test
    public void testIsEmpty() {
        Queue<Integer> queue = new Queue<>(3);
        assertTrue(queue.isEmpty());
        queue.addLast(1);
        assertFalse(queue.isEmpty());
    }

    @Test
    public void testSize() {
        Queue<Integer> queue = new Queue<>(5);
        queue.addLast(1);
        queue.addLast(2);
        queue.addLast(3);

        assertEquals(3, queue.size());
        queue.removeFirst();
        assertEquals(2, queue.size());
    }

}

File: src/test/queue/python/queue_test.py
import unittest
from src.main.queue.python.queue import Queue

class QueueTest(unittest.TestCase):

    def test_initialization(self):
        q = Queue(5)
        self.assertEqual(q.capacity, 5)
        self.assertEqual(q.size, 0)
        self.assertTrue(q.is_empty())
        self.assertFalse(q.is_full())
        print('Testes de inicialização da fila passaram!\n')

    def test_add_last(self):
        q = Queue(3)
        q.add_last(10)
        self.assertEqual(q.get_last(), 10)
        q.add_last(20)
        self.assertEqual(q.get_last(), 20)
        q.add_last(30)
        self.assertEqual(q.get_last(), 30)
        self.assertEqual(q.size, 3)
        self.assertTrue(q.is_full())
        with self.assertRaises(Exception): q.add_last(40)
        self.assertEqual(q.get_last(), 30)
        print('Testes de adição no final da fila passaram!')

    def test_add_first(self):
        q = Queue(3)
        q.add_first(10)
        self.assertEqual(q.get_first(), 10)
        q.add_first(20)
        self.assertEqual(q.get_first(), 20)
        q.add_first(30)
        self.assertEqual(q.size, 3)
        self.assertEqual(q.get_first(), 30)
        self.assertEqual(q.get_last(), 10)
        self.assertTrue(q.is_full())
        with self.assertRaises(Exception): q.add_first(40)
        self.assertEqual(q.get_first(), 30)
        print('Testes de adição no começo da fila passaram!')

    def test_add(self):
        q = Queue(5)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        q.add(15, 1)  # Adiciona 15 na posição 1
        self.assertEqual(q.size, 4)
        self.assertEqual(q.get(1), 15)
        self.assertEqual(q.get(2), 20)
        with self.assertRaises(Exception): q.add(40, 5)
        print('Testes de adição em índice determinado da fila passaram!')

    def test_remove_first(self):
        q = Queue(3)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        self.assertEqual(q.remove_first(), 10)
        self.assertEqual(q.size, 2)
        self.assertEqual(q.remove_first(), 20)
        self.assertEqual(q.remove_first(), 30)
        self.assertTrue(q.is_empty())
        with self.assertRaises(Exception): q.remove_first()
        print('Testes de remoção no começo da fila passaram!')

    def test_remove_last(self):
        q = Queue(3)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        self.assertEqual(q.remove_last(), 30)
        self.assertEqual(q.size, 2)
        self.assertEqual(q.remove_last(), 20)
        self.assertEqual(q.remove_last(), 10)
        self.assertTrue(q.is_empty())
        with self.assertRaises(Exception): q.remove_last()
        print('Testes de remoção no final da fila passaram!')

    def test_remove(self):
        q = Queue(5)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        q.add_last(40)
        self.assertEqual(q.remove(2), 30)
        self.assertEqual(q.size, 3)
        self.assertEqual(q.get(2), 40)
        with self.assertRaises(Exception): q.remove(5)
        print('Testes de remoção em determinado índice da fila passaram!')

    def test_get_first(self):
        q = Queue(3)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        self.assertEqual(q.get_first(), 10)
        q.remove_first()
        self.assertEqual(q.get_first(), 20)
        q.remove_first()
        self.assertEqual(q.get_first(), 30)
        q.remove_first()
        with self.assertRaises(Exception): q.get_first()
        print('Testes de busca no começo da fila passaram!')

    def test_get_last(self):
        q = Queue(3)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        self.assertEqual(q.get_last(), 30)
        q.remove_last()
        self.assertEqual(q.get_last(), 20)
        q.remove_last()
        self.assertEqual(q.get_last(), 10)
        q.remove_last()
        with self.assertRaises(Exception): q.get_last()
        print('Testes de busca no final da fila passaram!')

    def test_get(self):
        q = Queue(5)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        q.add_last(40)
        self.assertEqual(q.get(0), 10)
        self.assertEqual(q.get(2), 30)
        self.assertEqual(q.get(3), 40)
        with self.assertRaises(Exception): q.get(5)
        print('Testes de busca em determinado índice da fila passaram!')

    def test_index_of(self):
        q = Queue(5)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        q.add_last(20)
        self.assertEqual(q.index_of(20), 1)
        self.assertEqual(q.index_of(30), 2)
        self.assertEqual(q.index_of(40), -1)
        print('Testes de busca por primeiro elemento passaram!')

    def test_last_index_of(self):
        q = Queue(5)
        q.add_last(10)
        q.add_last(20)
        q.add_last(30)
        q.add_last(20)
        self.assertEqual(q.last_index_of(20), 3)
        self.assertEqual(q.last_index_of(30), 2)
        self.assertEqual(q.last_index_of(40), -1)
        print('Testes de busca por último elemento passaram!')

if __name__ == '__main__':
    unittest.main()
